vortex-ipc/src/reader.rs

```diff
@@ -168,51 +170,77 @@ impl<'a, R: Read> StreamArrayReader<'a, R> {
             vortex_bail!("Indices must be positive")
         }
 
-        if self.row_offset != 0 {
-            vortex_bail!("Stream has already been (at least partially) consumed")
-        }
+        Ok(TakeIterator {
+            reader: self,
+            indices,
+            row_offset: 0,
+        })
+    }
+}
 
-        let mut chunks = Vec::new();
+pub struct TakeIterator<'a, R: Read> {
+    reader: StreamArrayReader<'a, R>,
+    indices: &'a Array<'a>,
+    row_offset: usize,
+}
 
-        let mut row_offset = self.row_offset;
+/// NB: if the StreamArrayReader expires without being fully-consumed by the underlying iterator,
+/// it will leave the underlying messages stream in an unreadable state and free up a mutable ref
+/// to that stream, allowing someone else to try to issue a read that will inevitably fail.
+/// For this reason, we force full consumption of the reader when it goes out of scope.
+impl<'a, R: Read> Drop for StreamArrayReader<'a, R> {
+    fn drop(&mut self) {
+        while self
+            .next()
+            .expect("Unexpected failure consuming StreamArrayReader in destructor")
+            .is_some()
+        {
+            // Continue until the reader is exhausted
+        }
+    }
+}
+
+/// NB: if the TakeIterator expires without being fully-consumed by the underlying iterator,
+/// it will trigger the Drop impl on the StreamArrayReader, which will consume the rest of the
+/// messages for that array. This is necessary to ensure that the underlying stream is in a
+/// consistent state when the StreamArrayReader is dropped, but it also is not free. If users wish
+/// to avoid this work happening at exipry, users can just consume the rest of the iterator
+/// themselves when they see fit.
+impl<'a, R: Read> Iterator for TakeIterator<'a, R> {
```

Comment:
Let's use this: https://docs.rs/fallible-iterator/latest/fallible_iterator/trait.FallibleIterator.html

It will let us return `VortexResult<Option<Self::Item>>` and therefore you won't panic with all your expect/unwrap calls.